proposals/infra/INF-0004-register-types-and-diagnostics.md

```diff
@@ -0,0 +1,484 @@
+which is not allowed, so `err_hlsl_register_annotation_on_member` is emitted. 
+Next, some decls may have multiple register annotations applied. Regardless of the decl type,
+there will be validation that any pair of annotation statements may not have the same register
+type but different register number. If this is detected, `err_hlsl_conflicting_register_annotations`
+will be emitted.
+
+If the `resource` flag is set, check the register type against the resource class flag.
+If they do not match, `err_hlsl_mismatching_register_type_and_resource_type` is emitted.
+If the `basic` flag is set, then we first check if `default_globals` is set. If so, then 
+we check the register type. If the register type is 'b' or 'i', then
+`warn_hlsl_deprecated_register_type_b` or `warn_hlsl_deprecated_register_type_i` 
+will be emitted respectively, and treated as errors by default. If the errors are silenced,
+the compiler will function as if `c` was passed, and allocate the variables into the 
+global constant buffer.
+If the register type is 'c', it is allowed, but if `default_globals` is not set,
+`warn_hlsl_register_type_c_not_in_global_scope` will be emitted, 
+and will be treated as an error by default.
+
+After this point, `default_globals` doesn't need to be set.
+If 't', 'u', or 's' are given when `basic` is set, then
+`err_hlsl_mismatching_register_type_and_variable_type` will be emitted.
+If any other register type is seen, `err_hlsl_unsupported_register_type_and_variable_type`
+will be emitted instead.
+
+Finally, in the case that `udt` is set, we may have multiple register annotations
+applied to the decl. For every register type that is used to bind the resources contained 
+in the given UDT, we verify that the corresponding resource class flag has been set. These
+are the register types `t`, `u`, `b`, and `s`. If the corresponding resource class flag is 
+not set, `warn_hlsl_UDT_missing_resource_type_member` will be emitted.
+If `c` is given, then we must check that `contains_numeric` is set, and if not, 
+`warn_hlsl_UDT_missing_basic_type` is emitted.
+Otherwise, if any other register type is given, 
+then we emit `err_hlsl_unsupported_register_type_and_variable_type`.
+
+All the warnings introduced in this spec were not emitted in legacy versions of the compiler.
+The warnings, `warn_hlsl_register_type_c_not_in_global_scope`,
+`warn_hlsl_deprecated_register_type_b`,
+`warn_hlsl_deprecated_register_type_i`,
+`warn_hlsl_UDT_missing_resource_type_member`, and
+`warn_hlsl_UDT_missing_basic_type`, are all within a warning group known as 
+`disallow-legacy-binding-rules`. However, only `warn_hlsl_UDT_missing_resource_type_member`
+and  `warn_hlsl_UDT_missing_basic_type` are not treated as errors by default, 
+the rest of the warnings are errors by default. If legacy behavior is desired, a 
+user can pass the `--Wno-disallow-legacy-binding-rules` flag to the compiler to silence the

```

Comment:
Yes, here, legacy means earlier versions of DXC.
And yes, I think it wouldn't hurt to make sure that flag works on a `clang-dxc` invocation. Because I can define warning groups in clang, I would expect the option can be easily carried forward.

Reply:
No further work is necessary, I've confirmed that this flag works on clang-dxc.exe.
// RUN: %clang_dxc -T lib_6_3 -Vd -fsyntax-only -Wno-disallow-legacy-binding-rules %s
runs fine on `clang\test\SemaHLSL\resource_binding_attr_error_silence_diags.hlsl`